### DISCUSSION TITLE
Use of getOriginal() on model in observer doesn't return original value, when using transactions and ShouldHandleEventsAfterCommit.

### DISCUSSION THREAD

**Discussion by emil-nasso** [laravel/framework#52552] [Q&A]
### Laravel Version

11.20

### PHP Version

8.3

### Database Driver & Version

sqlite

### Description

When using and observer to subscribe to `updated` events on models, you can access the original value in a few different ways, including the `getOriginal(...)` method.

If the observer implements the `ShouldHandleEventsAfterCommit` and the changes to the model is wrapped in a databas transaction, the events should be processed after the transaction has commited. This works as expected.

However. When using transactions and `ShouldHandleEventsAfterCommit` the `getOriginal(...)` method no longer returns the original value.

I did not expect this, and can't find anything about it in the documentation. Is this a bug?

### Steps To Reproduce

Below are test cases for this. It consists of:

* A simple model.
* An observer that observes the model and, on `updated`, asserts that the value of `$model->test` is `updated` and the value of `$model->getOriginal('test')` is `original`
* Two test-cases, one with transactions and one without. The cases both create a model with the field `test` set to `original` and then updates that field to `updated`.

I would expect both of these test cases to pass, but the test case that uses transactions fails.

```php
<?php

namespace Tests;

use Illuminate\Database\Eloquent\Model;

class Test extends Model
{
    protected $connection = 'sqlite';

    protected $table = 'test';

    public $timestamps = false;

    protected $fillable = ['test'];
}
```

```php
<?php

namespace Tests;

use Illuminate\Contracts\Events\ShouldHandleEventsAfterCommit;
use PHPUnit\Framework\Assert;

class Observer implements ShouldHandleEventsAfterCommit
{
    public function updated(Test $model)
    {
        Assert::assertEquals('updated', $model->test);
        Assert::assertEquals('original', $model->getOriginal('test'));
    }
}
```

```php
<?php

namespace Tests;

use Illuminate\Database\Schema\Blueprint;
use Illuminate\Foundation\Testing\RefreshDatabase;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Schema;

class EventsInTransactionsTest extends TestCase
{
    use RefreshDatabase;

    protected function config(): void
    {
        config()->set('database.connections.sqlite', [
            'driver' => 'sqlite',
            'database' => ':memory:',
            'prefix' => '',
        ]);

        Schema::connection('sqlite')->create('test', function (Blueprint $table) {
            $table->id();
            $table->string('test');
        });

        Test::observe(Observer::class);
    }

    public function test_without_transaction(): void
    {
        $this->config();

        $model = new Test();
        $model->test = 'original';
        $model->save();

        $model->test = 'updated';
        $model->save();
    }

    public function test_with_transaction(): void
    {
        $this->config();

        DB::connection('sqlite')->transaction(function () {
            $model = new Test();
            $model->test = 'original';
            $model->save();

            $model->test = 'updated';
            $model->save();
        });
    }
}
```

**Comment by JakubSzczesniak** [comment]
It seems that you're dealing with a situation where the model object being tracked within a transaction is modified, which affects the behavior of the callbacks assigned to that transaction. This can lead to issues, especially when working with models in the context of a transaction, particularly in conjunction with events like updated.

In short, when a transaction is registered, Laravel creates callbacks that will be executed after the transaction is committed. If the model is modified during the transaction and Laravel uses the same object for the registered callbacks, this can lead to a scenario where the callbacks operate on outdated or inconsistent data.

In PHP, objects are passed by reference, meaning that if you modify an object in one place, those changes will be reflected everywhere that object is used unless you specifically clone the object before making changes.

**Comment by yurik94** [comment]
This is a big issue when building a logging mechanism for registering model changes, imagine this pretty common example:


Model to watch
```php
class Order extends Model{
   protected $fillable = [
        'state_id', // ...
   ];
}
```

Model to log changes
```php
#[ObservedBy(OrderObserver::class)]
class OrderStates extends Model{
   protected $fillable = [
        'order_id', 'old_state_id', 'new_state_id', // ...
   ];
   
  public function order(): BelongsTo
    {
        return $this->belongsTo(Order::class, 'order_id', 'id');
    }
}
```

Observer code
```php
class OrderObserver implements ShouldHandleEventsAfterCommit
{
    public function updated(Order $model) // tried also updating with no luck
    {
        $new_state_id = $model->state_id;
        $previous_state_id = $model->getOriginal('state_id'));
        
        // here I would like to log the change in a OrderStates entry       
        // but real previous state_id is lost when wrapping the change in a transaction...
    }
}
```

Example code
```php
DB::beginTransaction;

$order = Order::find(1);
$order->update(['state_id' => 2]);

// other relevant code...

DB::commit();
```

### Is there any other clean way to do this?

**Comment by macropay-solutions** [comment]
Solution
1. Put this behavior in documentation ([if not done already)](https://laravel.com/docs/11.x/eloquent#observers-and-database-transactions). This seems to be an issue converted into a discussion. Sad...  
2. Stop using **ShouldHandleEventsAfterCommit** and use queue with delay. In that job check if the model has the new state and if so use the old state (including timestamp) sent in the job to notify or log things @yurik94 . Do not use model id as job payload. Use model to array.

Thank you for this. We had thoughts of using this **ShouldHandleEventsAfterCommit** but we never used it. Lucky us.

**Reply by macropay-solutions** [reply]
Better solution https://github.com/laravel/framework/discussions/52552#discussioncomment-15271648

**Comment by AHMED-GAMAL-AG** [comment]
I have the same issue. Are there any updates?

**Comment by JaredMezz2** [comment]
I believe I've found a working solution for this! Avoid using ShouldHandleEventsAfterCommit and instead wrap any observer actions within manual DB::afterCommit() callbacks. 

Using your example, modify the provided Observer like the following;

```PHP
<?php

namespace Tests;

use Illuminate\Support\Facades\DB;
use PHPUnit\Framework\Assert;

class Observer
{
    public function updated(Test $model)
    {
        $original = $model->getOriginal('test');
        DB::afterCommit(function () use ($model, $original) {
            Assert::assertEquals('updated', $model->test);
            Assert::assertEquals('original', $original);
        });
    }
}
```

This maintains proper event firing order (saving --> updating --> SQL executed --> updated) for both transactions & non-transactions (these will be executed immediately) and retains original model values, all while ensuring that any external actions still only happen after the SQL is committed to the database.

I was building something similar to what @yurik94 had suggested, an outbox-style logging system that was triggered from both transactions and non-transactions within my codebase. Interestingly (or frustratingly depending on how you look at it), assuming your observer is firing from a transaction AND you're only making DB changes, since the observer is ran on the same PHP thread the transaction rollback **will properly undo any db changes**. This flow retains atomicity, however if the observer can also be fired from outside of a transaction and the changes fail to commit to your database you will introduce ghost actions that shouldn't have occurred, breaking the atomicity.

By not using ShouldHandleEventsAfterCommit and instead using manual DB::afterCommit we are able to have the best of both worlds, atomicity on the save and atomicity on the executed observer actions. 

Here's a real world example;

**Situation:** When a user's address is updated I want to send an email to them indicating changes were made while including the original address in the email.

```PHP
<?php

namespace App\Observers;

use App\Models\User;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Mail;
use App\Mail\AddressChangedEmail;

class Observer
{
    public function updated(User $user)
    {

        if ($user->wasChanged('address')) {
            $originalAddress = $user->getOriginal('address');

            DB::afterCommit(function() use ($user, $originalAddress) {
                Mail::to([$user->email])->send(new AddressChangedEmail($user, $originalAddress));
            });
        }
    }
}
```

This solved the problem for me!

**Reply by marius-ciclistu** [reply]
So that ShouldHandleEventsAfterCommit is useless?

**Reply by JaredMezz2** [reply]
@marius-ciclistu If you are using Database Transactions and need access to the original pre-change values, yes the ShouldHandleEventsAfterCommit will break getOriginal calls and be useless for you.

If you are not using transactions, ShouldHandleEventsAfterCommit can be used instead of these custom DB::afterCommit calls since by directly updating values through the Eloquent model your original values will be stored properly. 

If you are using transactions but **do not need** original values, you can still use ShouldHandleEventsAfterCommit.

I can't tell you exactly why a non-transaction update retains history when using the trait though, I'm assuming some Eloquent / Laravel handling will temporarily store those original values whereas when routed through a transaction the referenced PHP object has already been updated, saved, and flushed of old values. 

Another interesting point that threw me for a loop, if you're using ShouldHandleEventsAfterCommit and it's triggered by a database transaction `$model->wasChanged('address')` will properly return true, however the original values (accessed through `$model->getOriginal('address')`) only returns the updated value. So somehow the change detection is captured but those original field references have been cleared.

Fun stuff!

**Reply by marius-ciclistu** [reply]
That is related to model save function logic. Nothing fancy about wasChanged.

Still, saving, updating, deleting, creating called after commit makes no sense.

**Reply by marius-ciclistu** [reply]
Maybe this ShouldHandleEventsAfterCommit should affect only ed postfixed events and not ing postfixed ones.

**Reply by macropay-solutions** [reply]
<img width="1901" height="654" alt="image" src="https://github.com/user-attachments/assets/03e8fb43-a2f3-41e8-ae5b-f6d7841772d9" />
Maybe this would fix it.
https://github.com/macropay-solutions/maravel-framework/pull/25

**Reply by JaredMezz2** [reply]
> The ed postfixed events make sense and in those, original is not to be used, but wasChanged should.

What about a scenario like the example I provided, needing to perform some external action after the update has completed and persisted in the database while including the original data? Whether it be a queued job, an email being sent, etc. etc.

I am under the impression updating() is moreso for direct repercussions from a value change on the same model instance (ie if address changed, automatically append a timestamp to address_changed_time field on the same User model) whereas updated() will handle the email being delivered (completely external from the model but still needs to happen every time on address change).

> Maybe this ShouldHandleEventsAfterCommit should affect only ed postfixed events and not ing postfixed ones.

I think I agree with this, you're right in saying that it doesn't make a ton of sense that saving/updating/deleting is fired after the data is committed. That being said by ensuring all happen after the commit at least the trait is consistent and doesn't include caveats surrounding the type of event.

If they wanted to keep it as is, perhaps a second ShouldExecuteEventsAfterCommit trait could be added that automatically wraps all functionality within these event hooks in DB::afterCommit calls instead of the entire model lifecycle being triggered after the commit.

**Reply by marius-ciclistu** [reply]
Now I see that the syncOriginal is called only after the updated event is fired for example so you could use the original also in updated event. syncChanges is called just before the updated event so you can use that to check if a column was changed.

>ShouldExecuteEventsAfterCommit

 see ShouldDispatchAfterCommit

I will edit my above reply about original in -ed event.

**Reply by macropay-solutions** [reply]
The interesting thing would be to preserve somehow the original for the -ed events...

**Reply by JaredMezz2** [reply]
@marius-ciclistu ShouldDispatchAfterCommit is an interesting one. Logging-wise this acts exactly as I'd want ShouldHandleEventsAfterCommit to, 

Saving --> Updating --> SQL Executed --> Updated --> Saved, all while retaining the original values within the updated() step.

That being said it does not (on my current Laravel 10 version at least) only execute the updated() content after the data is committed to the database. In my provided example the email is still sent out despite an exception preventing the DB save, however it's a step in the right direction since I do indeed have access to the old address.

**Reply by macropay-solutions** [reply]
@JaredMezz2 ShouldDispatchAfterCommit must be used for event classes not for listeners. That is why you get all events executed before commit.

However this ShouldHandleEventsAfterCommit is affecting -ing events which makes no sense. Your example with DB::afterCommit makes more sense to be used instead of this ShouldHandleEventsAfterCommit, making it useless...

**Reply by macropay-solutions** [reply]
In the docs created is used as example https://laravel.com/docs/12.x/eloquent#observers-and-database-transactions so maybe that is the only situation in which ShouldHandleEventsAfterCommit should be used. But still, it will affect the creating, saving events... So maybe [our PR](https://github.com/macropay-solutions/maravel-framework/pull/25) makes sense to be merged...

**Reply by macropay-solutions** [reply]
@JaredMezz2 Thank you https://github.com/macropay-solutions/maravel-framework/pull/25/files

**Reply by JaredMezz2** [reply]
@macropay-solutions Awesome, glad I could be of some help! ShouldDispatchAfterCommit does indeed make more sense for Event classes.

Your solution looks very clean, I'm a fan. I still unfortunately run into the same issue this thread was originally about (->getOriginal() returning updated values) unless I completely step away from Observers so I did whip up this solution that seems to do the trick.

https://github.com/laravel/framework/pull/58149

I think the correct answer probably lies somewhere between our two pulls.. I'm not a huge fan of creating copies for all models passed into the -ed methods however this seemed to be the only way to ensure 1) actions are ran after the changes have been committed while using transactions and 2) we can actually access the original values in those methods. That is, without the manually added DB::afterCommit() calls.

Glad to see this thread wasn't a lost cause!

**Reply by marius-ciclistu** [reply]
I am curious if your PR to Laravel 12 will get merged. What if the model is updated twice?

Update 1.

Update other tables

Update 2

Then which originals will you get in the observer updated?

**Reply by JaredMezz2** [reply]
I'm also curious!

Using address as an example again and assuming I'm changing it from 123 Home St -> 321 Prince St -> 999 Main St

If the model is updated twice within the same transaction, I see the log history as the following;

1. SQL update to "321 Prince St" executed (1)
2. SQL update to "999 Main St" executed (2) 
3. Saving (1) 
4. Updating (1)
5. Updated (1, original == "123 Home St ")
6. saved (1) 
7. saving (2)
8. updating (2)
9. updated (2, original == "321 Prince St") 
10. saved (2).

If the model is updated twice in unique transactions, the log path follows my initial example (SQL -> updating -> updated) and getOriginal values track as expected.

If the model is updated twice without transactions, original values are as expected.

Did you have a different model change path in mind? I'd be happy to test that out as well. This copy _should_ be getting made immediately upon invocation, therefore allowing it to copy at-time values before the event is even "queued" up to be processed once the commit has finished. Then the same PHP passing by reference functionality that was preventing the use of `->getOriginal()` initially is now allowing us to keep snapshots of this model throughout it's transaction lifecycle and pass those along to the eventually ran -ed methods.

**Reply by marius-ciclistu** [reply]
Nice. The 2 new setters that you added to be used only for the cloned models could be used for any model. Maybe you should add just one new methot that will return the cloned model. Did you try to just clone the model?

**Reply by marius-ciclistu** [reply]
snapshotModelForAfterCommit can be moved in the model and so you can remove the 2 new setters. In that way you avoid polution of the model with dangerous public methods.

**Reply by marius-ciclistu** [reply]
But still any relation from the model will not retain its state in first update because it is an object. 

Begin
Update User
Update Relation
Update User
Commit

The relation will have same state in update 1 AND update 2 because it is an object.

I still think your first solution with DB:afterCommit is more bulletproof because you can snapshot the values that you are interested at the right moment.  Your PR just guesses what values you would need. Plus, might be slower than your first solution because you are doing extra stuff that might not be used.

**Reply by marius-ciclistu** [reply]
in_array should always be used with third param true. I forgot yesterday.

\in_array(0, ['a']) is true in 7.2 and false in 8.2

**Reply by JaredMezz2** [reply]
Thank you @marius-ciclistu, this back and forth is super helpful!

It does almost look like simply returning the $clone model without running all those methodical setters still returns the same object, I'll double check a few more things and most likely just return the direct clone instead. That negates any worry about public setters on those models! I mostly set those up to maintain consistency with the other getters / setters, some of the available options were already direct overwrites so _safety_  didn't seem like the biggest priority.

Your relation state is an interesting point. You're correct that the state is identical when the second round of observer methods fire when inside of the same transaction, but it looks like the state is stale even if we're performing these updates outside of a transaction.

Something like the following still provides me the old relational data;

```PHP
$user = User::find(1);
$user->update(['address' => '321 Prince St']);
// Observer fired now

// Just manually updating a related record by id, 'originalValue' to 'newValue'
UserRelatedModel::where('id', 1)->update(['someKey' => 'newValue']);

$user->update(['address' => '999 Main St']);
// Observer fired for the second time

```

I wonder if the model needs to be rehydrated or something before it would be seen, but perhaps that's just not how observers should be used. I'm watching for updates on the User model, I shouldn't be basing my decisions within the observer methods on a related user model.

Good call on the strict type checking for in_array!

I also do really feel that DB::afterCommit is the cleanest; not having to duplicate unnecessary fields that won't be used and it keeps all events firing in the proper order. I will most likely continue using this method, however that's mainly because my projects on Laravel 10 without any plans to upgrade to 12 in the near future. I'm just not super satisfied with how this trait inconsistently impacts available observer methods!

**Reply by marius-ciclistu** [reply]
$user->userRelated->update([...]); 

This is how you should update the relation. Taylor closed your PR.

**Reply by marius-ciclistu** [reply]
@JaredMezz2 
>Laravel 10 without any plans to upgrade to 12 in the near future. I'm just not super satisfied with how this trait inconsistently impacts available observer methods!

It is an interface not a trait. I fell the same way about the changes from 11. That is why I offered an alternative to 10.

**Reply by JaredMezz2** [reply]
Hah, replacing with `$user->userRelated->update([])` now shows me the updated value in both logs. Pretty much just as useless! Seems like I'll just be staying away from ShouldHandleEventsAfterCommit and routing through DB::afterCommit in my -ed methods.

Taylor's event suggestion makes sense however still leaves this interface implementation inconsistent, it is what it is! In an ideal world I should most likely be routing all address updates through a single method (or setter, or something I'm unaware of) that way I can handle any changes in a single spot that's not an observer.

& yes interface, my bad! Thanks again for all the chat @marius-ciclistu @macropay-solutions

**Reply by marius-ciclistu** [reply]
Bloating your code base with N events when you can handle all in an observer is definitelly not my first choice.

**Reply by macropay-solutions** [reply]
@JaredMezz2 You gave us another idea to optimize the observers.
<img width="1900" height="605" alt="image" src="https://github.com/user-attachments/assets/a2dce116-d93b-4599-b1bc-76b93f337232" />

**Reply by JaredMezz2** [reply]
@macropay-solutions That's interesting! For the base case of the default Observer event methods that is probably a little quicker, especially if you're not utilizing every available event.

Could this change potentially be less-efficient if I have several helper methods within my observer though? Ie if I had normalizeFoo(), buildPayload(), etc. etc, would these now be getting caught by the get_class_methods, loaded into memory, and result in a slightly more expensive comparison? With the original code these helper methods would never be checked.

Perhaps they'd be excluded with proper private scoping, I'm not sure whereabouts this registerObserver actually _lives_ and what it can see.

**Reply by macropay-solutions** [reply]
The array intersect eliminates your concern. That is better than calling method exists N times for each observer at boot time.

We are looking into event:cache now so that it can autodiscover also observers so that Model::observe does not need to be called.

**Reply by macropay-solutions** [reply]
@JaredMezz2 have a look https://github.com/macropay-solutions/maravel-framework/pull/26/files
Update. We managed to improve also the registration of events and cached observers.

We released these improvements under version [10.56.0](https://github.com/macropay-solutions/maravel-framework/releases/tag/10.56.0)

**Reply by macropay-solutions** [reply]
Maybe Laravel should adopt this improvement also for caching Observers and Events and also improving their registration.
 Needless to say that the subscribers are not very fast at boot time because they need to be instantiated and subscribe method needs to be called in them for the dispatcher listen method to be called...
